sycl/source/detail/scheduler/graph_builder.cpp

```diff
@@ -1387,6 +1481,186 @@ Command *Scheduler::GraphBuilder::connectDepEvent(
+  // the list will be added in a later PR.
+  auto FusedCG = nullptr;
+
+  if (!FusedCG) {
+    // If the JIT compiler returns a nullptr, JIT compilation of the fused
+    // kernel failed. In that case, simply cancel the fusion and run each kernel
+    // on its own.
+    auto LastEvent = PlaceholderCmd->getEvent();
+    this->cancelFusion(Queue, ToEnqueue);
+    return LastEvent;
+  }
+
+  // Remove internal explicit dependencies, i.e., explicit dependencies from one
+  // kernel in the fusion list to another kernel also in the fusion list.
+  auto &FusedEventDeps = FusedCG->MEvents;
+  FusedEventDeps.erase(
+      std::remove_if(
+          FusedEventDeps.begin(), FusedEventDeps.end(),
+          [&](EventImplPtr &E) { return E->getCommand() == PlaceholderCmd; }),
+      FusedEventDeps.end());
+
+  auto FusedKernelCmd =
+      std::make_unique<ExecCGCommand>(std::move(FusedCG), Queue);
+
+  assert(PlaceholderCmd->MDeps.empty());
+  // Next, backwards iterate over all the commands in the fusion list and remove
+  // them from the graph to restore the state before starting fusion, so we can
+  // add the fused kernel to the graph in the next step.
+  // Clean up the old commands after successfully fusing them.
+  for (auto OldCmd = CmdList.rbegin(); OldCmd != CmdList.rend(); ++OldCmd) {
+    removeNodeFromGraph(*OldCmd, ToEnqueue);
+    cleanupCommand(*OldCmd, /* AllowUnsubmitted */ true);
+  }
+
+  createGraphForCommand(FusedKernelCmd.get(), FusedKernelCmd->getCG(), false,
+                        FusedKernelCmd->getCG().MRequirements,
+                        FusedKernelCmd->getCG().MEvents, Queue, ToEnqueue);
+
+  ToEnqueue.push_back(FusedKernelCmd.get());
+
+  std::vector<Command *> ToCleanUp;
+  // Make the placeholder command depend on the execution of the fused kernel
+  auto *ConnectToPlaceholder =
+      PlaceholderCmd->addDep(FusedKernelCmd->getEvent(), ToCleanUp);
+  if (ConnectToPlaceholder) {
+    ToEnqueue.push_back(ConnectToPlaceholder);
+  }
+  for (Command *Cmd : ToCleanUp) {
+    cleanupCommand(Cmd);
+  }
+  ToEnqueue.push_back(PlaceholderCmd);
+
+  if (MPrintOptionsArray[AfterFusionComplete]) {
+    printGraphAsDot("after_fusionComplete");
+  }
+
+  // Set the status for the fusion command.
+  PlaceholderCmd->setFusionStatus(KernelFusionCommand::FusionStatus::COMPLETE);
+
+  return FusedKernelCmd.release()->getEvent();
+#else  // SYCL_EXT_CODEPLAY_KERNEL_FUSION
+  printFusionWarning("Kernel fusion not supported by this build");
+  (void)PropList;
+  auto FusionList = findFusionList(QUniqueID);
+  auto *PlaceholderCmd = FusionList->second.get();
+  auto LastEvent = PlaceholderCmd->getEvent();
+  this->cancelFusion(Queue, ToEnqueue);
+  return LastEvent;
+#endif // SYCL_EXT_CODEPLAY_KERNEL_FUSION
+}
+
+bool Scheduler::GraphBuilder::isInFusionMode(QueueIdT Id) {
+  auto FusionList = findFusionList(Id);
+  if (FusionList == MFusionMap.end()) {
+    return false;
+  }
+  return FusionList->second->isActive();
+}
```

Comment:
For the sake of keeping the interface somewhat consistent, would it make sense to have a

```suggestion
}

bool Scheduler::GraphBuilder::isInFusionMode(QueueImplPtr Queue) {
  return isInFusionMode(std::hash<QueueImplPtr>()(Queue));
}
```

and use that in `queue_impl`?

Reply:
I see your point about the interface. However, `QueueImplPtr` is `std::shared_ptr<detail::queue_impl>`, so in order to be able to call this from `queue_impl`, we would need to create a new shared pointer from `this` in `queue_impl`, which would lose the reference count.

Reply:
Good point. Let's keep it as-is.